DataDonationCryptoObjC/DataDonationCryptoObjC/Sources/InternalServices/DataDonationCryptor.swift

```diff
@@ -0,0 +1,77 @@
+//  Copyright (c) 2021 D4L data4life gGmbH
+//  All rights reserved.
+//
+//  D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
+//  including any intellectual property rights that subsist in the SDK.
+//
+//  The SDK and its documentation may be accessed and used for viewing/review purposes only.
+//  Any usage of the SDK for other purposes, including usage for the development of
+//  applications/third-party applications shall require the conclusion of a license agreement
+//  between you and D4L.
+//
+//  If you are interested in licensing the SDK for your own applications/third-party
+//  applications and/or if you’d like to contribute to the development of the SDK, please
+//  contact D4L by email to help@data4life.care.
+//
+
+import Foundation
+import Data4LifeCrypto
+import CryptoKit
+
+final class DataDonationCryptor {
+
+    private let donorKeyHolder: DonorKeyHolderProtocol
+    private let ivGenerator: InitializationVectorGeneratorProtocol
+    private let coreCryptoService: CoreCryptoServiceProtocol
+
+    private let programName: String
+    private let encryptionVersion: UInt8
+
+    init(programName: String,
+         encryptionVersion: UInt8 = 2,
+         donorKeyHolder: DonorKeyHolderProtocol = DonorKeyHolder(),
+         ivGenerator: InitializationVectorGeneratorProtocol = InitializationVectorGenerator(),
+         coreCryptoService: CoreCryptoServiceProtocol = CoreCryptoService()) {
+        
+        self.donorKeyHolder = donorKeyHolder
+        self.ivGenerator = ivGenerator
+        self.coreCryptoService = coreCryptoService
+
+        self.programName = programName
+        self.encryptionVersion = encryptionVersion
+    }
+}
+
+extension DataDonationCryptor: DataDonationCryptorProtocol {
+
+    static func make(programName: String) -> DataDonationCryptorProtocol {
+        return DataDonationCryptor(programName: programName)
+    }
+
+    func encrypt(_ plainBody: Data) throws -> Data {
```

Comment:
i kinda postponed that decision to later because i wasnt sure if we were actually using it.
As it is now it wouldnt be for every ecnryption round but just for every encryption keypair (which would mean we would need an ALP or a Donation one), so  i think i should inject it also for being coherent with the keyholder, if you agree

Reply:
i kinda postponed that decision to later because i wasnt sure if we were actually using it.
As it is now it wouldnt be for every ecnryption round but just for every encryption keypair (which would mean we would need an ALP or a Donation one), so  i think i should inject it also for being coherent with the keyholder, if you agree

Reply:
Well, please keep in mind, we are using not the DonorKeyPair...we are just have the DonationService PublicKey and the one of ALP for encryption. Also we need to encrypt in one Round **with both of them** (see [here](https://github.com/gesundheitscloud/data-donation-sdk-js/blob/02929ffb14a17257356fe5188d6a15702b823d8c/src/sdk.ts#L300) and [here](https://github.com/gesundheitscloud/data-donation-sdk-js/blob/02929ffb14a17257356fe5188d6a15702b823d8c/src/sdk.ts#L284)).
The DonorKey is only for signing, which is also in the mix while doing rounds.

Reply:
Did changes reflecting that, i confused the donorService keys for the Donor keys from the js code thats why i thought it wasnt needed, my bad. please double check if new changes could work